core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java

```diff
@@ -31,36 +36,113 @@
 
     static final FilteredQueryBuilder PROTOTYPE = new FilteredQueryBuilder(null, null);
 
     /**
+     * Returns a {@link MatchAllQueryBuilder} instance that will be used as
+     * default queryBuilder if none is supplied by the user. Feel free to
+     * set queryName and boost on that that instance - it's always a new one.
+     * */
+    public static QueryBuilder generateDefaultQuery() {
+        return new MatchAllQueryBuilder();
+    }
+
+    /**
+     * A query that applies a filter to the results of a match_all query.
+     * @param filterBuilder The filter to apply on the query (Can be null)
+     * */
+    public FilteredQueryBuilder(QueryBuilder filterBuilder) {
+        this(generateDefaultQuery(), filterBuilder);
+    }
+
+    /**
      * A query that applies a filter to the results of another query.
      *
-     * @param queryBuilder  The query to apply the filter to (Can be null)
+     * @param queryBuilder  The query to apply the filter to
      * @param filterBuilder The filter to apply on the query (Can be null)
      */
-    public FilteredQueryBuilder(@Nullable QueryBuilder queryBuilder, @Nullable QueryBuilder filterBuilder) {
-        this.queryBuilder = queryBuilder;
-        this.filterBuilder = filterBuilder;
+    public FilteredQueryBuilder(QueryBuilder queryBuilder, QueryBuilder filterBuilder) {
+        this.queryBuilder = (queryBuilder != null) ? queryBuilder : generateDefaultQuery();
+        this.filterBuilder = (filterBuilder != null) ? filterBuilder : EmptyQueryBuilder.PROTOTYPE;
+    }
+
+    /** Returns the query to apply the filter to. */
+    public QueryBuilder query() {
+        return queryBuilder;
+    }
+
+    /** Returns the filter to apply to the query results. */
+    public QueryBuilder filter() {
+        return filterBuilder;
     }
 
     @Override
-    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.startObject(NAME);
-        if (queryBuilder != null) {
-            builder.field("query");
-            queryBuilder.toXContent(builder, params);
+    protected boolean doEquals(FilteredQueryBuilder other) {
+        return Objects.equals(queryBuilder, other.queryBuilder) &&
+                Objects.equals(filterBuilder, other.filterBuilder);
+    }
+
+    @Override
+    public int doHashCode() {
+        return Objects.hash(queryBuilder, filterBuilder);
+    }
+
+    @Override
+    public Query doToQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
+        Query result = null;
+
+        Query query = queryBuilder.toQuery(parseContext);
+        Query filter = filterBuilder.toQuery(parseContext);
+
+        if (query == null) {
+            // Most likely this query was parsed from the JSON query DSL - it parsed to null so we ignore
+            // the whole filtered query as there is nothing to filter on. See FilteredQueryParser for an example.
+            return null;
```

Comment:
I think we have to clarify this comment, it didn't parse to null, that never happens anymore (othersie you would have NPE at line 101). This became a null lucene query after calling toQuery.